node-graph/README.md

```diff
@@ -54,34 +72,49 @@ pub fn multiply_opacity(document_node: &DocumentNode, node_id: NodeId, _context:
 
 ```rs
+use crate::{Color, Node};
+
 #[derive(Debug, Clone, Copy)]
-pub struct OpacityNode<O> {
-	opacity_multiplier: O,
+pub struct OpacityNode<OpacityMultiplierInput> {
+	opacity_multiplier: OpacityMultiplierInput,
 }
 
-impl<'i, N: Node<'i, (), Output = f64> + 'i> Node<'i, Color> for OpacityNode<N> {
+impl<'i, OpacityMultiplierInput: Node<'i, (), Output = f64> + 'i> Node<'i, Color> for OpacityNode<OpacityMultiplierInput> {
 	type Output = Color;
-	fn eval<'s: 'i>(&'s self, color: Color) -> Color {
+	fn eval(&'i self, color: Color) -> Color {
 		let opacity_multiplier = self.opacity_multiplier.eval(()) as f32 / 100.;
 		Color::from_rgbaf32_unchecked(color.r(), color.g(), color.b(), color.a() * opacity_multiplier)
 	}
 }
+```
 
-impl<N> OpacityNode<N> {
-	pub fn new(node: N) -> Self {
-		Self { opacity_multiplier: node }
-	}
+The `eval` function can only take one input. To support more than one input, the node struct can store references to other nodes. This can be seen here, as the `opacity_multiplier` field, which generic and is is constrained to the trait `Node<'i, (), Output = f64>`. This means that it is a node with the input of `()` (no input is required to comput the opacity) and an output of an `f64`.
+
+To compute the value when executing the `OpacityNode`, we need to call `self.opacity_multiplier.eval(())`. This evaluates the node that provides the `opacity_multiplier` input, with the input value of `()` - nothing. This occurs each time the opacity node is run.
+
+To test this:
+```rs
+#[test]
+fn test_opacity_node() {
+	let opacity_node = OpacityNode {
+		opacity_multiplier: crate::value::CopiedNode(10f64), // set opacity to 10%
+	};
+	assert_eq!(opacity_node.eval(Color::WHITE), Color::from_rgbaf32_unchecked(1., 1., 1., 0.1));
 }
 ```
 
-The `eval` function can only take one input. To support more than one input, the node struct can contain references to other nodes (it is the references that implement the `Node` trait). If the input is a constant, then it will reference a node that simply evaluates to a constant. If the input is a node, then the relevant proto-node will be referenced. To evaluate the opacity multiplier input, you can pass in `()` (because no input is required to calculate the opacity multiplier) which returns an `f64`. This is because of the generics we have applied: `N: Node<'i, (), Output = f64>` A helper function to create a new node struct is also defined here.
+The `graphene_core::value::CopiedNode` is a node that, when evaluated, copies `10f32` and returns it.
 
-This process can be made more concise using the `node_fn` macro, which can be applied to a function like `image_opacity` with an attribute of the name of the node:
+## Creating a new protonode
+
+Instead of manually implementing the `Node` trait with complex generics, one can use the `node_fn` macro, which can be applied to a function like `image_opacity` with an attribute of the name of the node:
```

Comment:
should we mention the other node_fn like macros as well here?

Reply:
`node_new` and `node_impl`?

Reply:
Yeah, `node_new` can be used if you need to specify the implementation manually and `node_impl` can be used if you want to add multiple impl blocks which would otherwise create conflicting new functions